### PR TITLE
Add support in hypes to load segmentation masks

### PR DESCRIPTION
**Pull request by MartinThoma**
Users can define in the hypes file on a high level which colors should
be which classes.

### CODE REVIEW

**Comment by MarvinTeichmann** [comment]
Overall it looks nice. However are you sure, that we should use HTML notation for color definition? Cityscape as well as Kitti Road both define segmentation classes by [RGB values](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py) (in the corresponding documentation files).  I do not know about the other DataSets,  but from a DataSet perspective it makes sense to provide the User with RGB values. Maybe it is possible to provide both options?

Additionally in `load_segmentation_mask` you are iterating over each pixel in python. Have you measured the performance of this? Without testing it, I would rather believe that this is very _very_ slow. Loading the segmentation mask happens at runtime. The loop might introduce a bottleneck in the training process.

Instead of iterating over each pixel it is possible to use numpy logical operations to archive the same. (Technically numpy will still iterate over all pixels, but it is implemented in speedy c potentially parallelized end certainly done a lot of optimization.) Find below my implementation of this, which I use to get the segmentation mask of cityscape (from RGB Image). The key lines are `gt_label = np.all(gt_image == label.color, axis=2)` were `label.color` is the RGB-Value (e.g. `[0,255,120]`) and `gt_reshaped = np.dot(gt_image, vectors)`.

Additionally does your code assume, that gt is RGB? This is the case for most data-sets. However it is possible to code classes in greyscale (e.g. MediSeg). Do you know a good way to generalize that?

```
def transform_dense_to_sparse(gt_image):
    """Transforms ground truth RGB to segmentation mask.

    Parameters
    ----------
    gt_image : RGB Image

    Returns
    -------
    segmentation_mask
    """

    shape = gt_image.shape
    gt_reshaped = np.empty([shape[0], shape[1], 19])
   # num_classes = 19 (for cityscape)

    for label in label_dict.labels:
        if label.trainId == 255 or label.trainId == -1:
            # skip classes which are ignored in the challenge
            continue

        gt_label = np.all(gt_image == label.color, axis=2)
        gt_reshaped[:, :, label.trainId] = gt_label

    return gt_reshaped


def transform_sparse_to_dense(gt_image):
    """Transforms segmentation mask to RGB.

    Parameters
    ----------
    gt_image : segmentation_mask

    Returns
    -------
    gt_reshaped: gt as RGB image
    """

    vectors = np.empty([19, 3])

    for label in label_dict.labels:
        if label.trainId == 255 or label.trainId == -1:
            # skip classes which are ignored in the challenge
            continue
        vectors[label.trainId, :] = label.color

    gt_reshaped = np.dot(gt_image, vectors)

    shape = gt_image.shape
    assert(gt_reshaped.shape == (shape[0], shape[1], 3))

    return gt_reshaped
```

**Comment by MartinThoma** [comment]
> are you sure, that we should use HTML notation for color definition? 

Pro HTML color codes:
- Easy to recognize as colors
- Easy to get the color from an image with most image editors
- It is relatively clear how the JSON file should look like

> Cityscape as well as Kitti Road both define segmentation classes by RGB values

Right. That is not a reason not to accept HTML color codes, but probably a reason to accept RGB lists (JSON does not know tuples, AFAIK). Currently, RGB tuples are possible (see https://github.com/TensorVision/TensorVision/pull/45/files#diff-a1830f8ae3b24127add7acc42ebbdb47R439)

**Comment by MartinThoma** [comment]
> Additionally does your code assume, that gt is RGB? This is the case for most data-sets. However it is possible to code classes in greyscale (e.g. MediSeg). Do you know a good way to generalize that?

Yes, I thought about it. [This line](https://github.com/TensorVision/TensorVision/pull/45/files#diff-a1830f8ae3b24127add7acc42ebbdb47R497) makes sure the data is loaded as RGB. If it is in grayscale, it converts it to RGB.

edit: I should probably document this. While this behavior makes a lot of sense, it might be surprising for the user.

edit2: It is documented.

**Comment by MartinThoma** [comment]
@MarvinTeichmann Now about the performance part: You are right. I did not measure the performance, but I'm also pretty sure using numpy operations is better. I fixed it.

**Comment by MartinThoma** [comment]
Hm, I'm currently thinking that it might make sense to return a tuple from `load_segmentation_mask`: The segmentation mask with the classes as well as a mask `ignore pixels` which has `True` at positions where the pixel should be ignored.

However, this is also possible to be done on user side. The user can check his `hypes['classes']` which index the `ignore` class has and thus easily generate this matrix by

```
ignore_pixels = (segmentation_mask == ignore_class_index)
```

So I think I'd rather keep the code as it is.

**Comment by MarvinTeichmann** [comment]
The Code looks good! One additional question, you are returning a dense Segmentation mask (i.e. shape [height, width] and classes as integer values). The current loss implementation however needs a sparse segmentation mask (shape [height, width, num_classes]). How are you going to utilize the return value? 

Other from that it is fine. You can go ahead and merge, if you like.

Edit: Regarding `ignore pixels`. I just gave it a thought and maybe it is indeed a good idea, to return a segmentation mask (without an `ignore_class_index`) and an `ignore pixels` mask. This will make it easy to use the segmentation mask in a loss function.

**Comment by MarvinTeichmann** [comment]
Ok, I gave it another thought. If you want to optimize softmax_cross entropy in tensorflow, the best possible way is to utilize an integer segmentation map, using [this](https://www.tensorflow.org/versions/r0.9/api_docs/python/nn.html#sparse_softmax_cross_entropy_with_logits) predefined loss function. 

However, if you want to implement your own cross_entropy ([which is necessary sometimes](https://github.com/tensorflow/tensorflow/issues/2327)), or you want to use a different loss (e.g. optimize MeanIoU directly).  One need a binary segmentation map.

**Comment by MarvinTeichmann** [comment]
But if the merge is urgent to you, you can also go ahead an merge. We can change the interface later.
